pySDC/projects/PinTSimE/battery_2capacitors_model.py

```diff
@@ -33,19 +66,43 @@ def run():
 
     problem_classes = [battery_n_capacitors]
     sweeper_classes = [imex_1st_order]
+    num_nodes = 4
+    restol = -1
+    maxiter = 12
```

Comment:
I'm also very confused here. I wanted to use a residual tolerance to reach a stop either the residual drops below the tolerance or after reaching `maxiter`. But setting e.g. `restol = 1e-15` leads to no iterations, i.e., nothing is done because `L.status.residual=0.0` is set somewhere.. Do I have to set something using this `skip_residual_computation`? But this does not make any sense, because the discontinuous test ODE problem works with setting an residual tolerance.. So confusing.. it feels like pySDC and everything else is hating me since yesterday..

Reply:
Hmmmm. Is this for an RK sweeper or something like that? The residual computations should not be skipped by default, but if you do skip it, it will be initialised to 0. Sounds like this is indeed the issue.
I hope pySDC and you can make up soon 🥴

Reply:
No, I don't want to skip that! This is why I am confused I want to use the residual tolerance. But it does not work. Probably the problem is sitting in front of the laptop.

Reply:
The problem is probably sitting in front of my laptop. What's the output when you decrease the logger level? What do you get for `skip_residual_computation` in the sweeper params and what does the code say about the residual if it is computing any?

Reply:
I set `logger_level=15` so it should output something. But there is not output. The initial output after initialization of the controller states `skip_residual_computation = ()`. I can see no computation. I was printing out `L.status.residual` in check_convergence.py and it is set to 0.0. Hence, no computation because we are "converged".

Reply:
The residual for both battery models is zero already before start of computation. The reason why nothing is happening.. Now I remember why I'm using `sweeper_params['initial_guess'] = 'zero'` when I started to implement the battery models one year ago. :D But this not really explain why there is no logging output, or maybe yes? ...

Reply:
I started a PR on your fork with a fix for this. The issue is not the initial guess, but how the `A` matrix is initialised. Did you not get an email about the PR?

Reply:
No email there.. Thanks for the PR!

Reply:
By the way: The built in python debugger proved to be quite useful in debugging this. @kyrillh was making fun of me until I finally gave it a try and it's actually pretty good! You can just write 
```
breakpoint()
```
in the code and when you execute it, the code will stop there. Now you see the output generated before and have the option to manipulate everything. You move on by doing `c` for continue. This is very handy when the code does many iterations of the same thing. If I just print, I have to scroll around like crazy in the end. If I do `raise` after the print as I used to, I can't see what happens in the next iteration. You can exit the debugger with `ctrl-D`
In this case, I printed the `A` matrix in `eval_f` and had a breakpoint in `compute_residual` then I noticed the matrix is 0 in the first iteration and non-zero later.
Maybe you know about this already, but I believe the built-in debugger is under-appreciated. I believed myself that debugging is reserved for fancy IDEs until very recently. But luckily there is no need to give in to IDEs ;)

Reply:
Hey, that's very cool! I just used that and it should be very helpful! Since now I used many print lines for debugging, very annoying.. But now I can do my life a bit easier with using an editor. :D